components/renku_data_services/crc/db.py

```diff
@@ -466,6 +487,62 @@ async def update_resource_class(
+            res_classes = await self.get_classes(api_user, class_id, resource_pool_id=resource_pool_id)
+            if len(res_classes) == 0:
+                raise errors.MissingResourceError(
+                    message=f"The resource pool with ID {resource_pool_id} or the resource "
+                    f"class with ID {class_id} do not exist, or they are not related."
+                )
+            stmt = select(schemas.TolerationORM).where(schemas.TolerationORM.resource_class_id == class_id)
+            res = await session.execute(stmt)
+            return [i.key for i in res.scalars().all()]
+
+    @_only_admins
+    async def delete_tolerations(self, api_user: base_models.APIUser, resource_pool_id: int, class_id: int):
+        """Delete all tolerations for a specific resource class."""
+        async with self.session_maker() as session, session.begin():
+            res_classes = await self.get_classes(api_user, class_id, resource_pool_id=resource_pool_id)
+            if len(res_classes) == 0:
+                raise errors.MissingResourceError(
+                    message=f"The resource pool with ID {resource_pool_id} or the resource "
+                    f"class with ID {class_id} do not exist, or they are not related."
+                )
+            stmt = delete(schemas.TolerationORM).where(schemas.TolerationORM.resource_class_id == class_id)
+            await session.execute(stmt)
+
+    @_only_admins
+    async def get_affinities(
+        self, api_user: base_models.APIUser, resource_pool_id: int, class_id: int
+    ) -> List[models.NodeAffinity]:
+        """Get all affinities for a resource class."""
+        async with self.session_maker() as session:
+            res_classes = await self.get_classes(api_user, class_id, resource_pool_id=resource_pool_id)
+            if len(res_classes) == 0:
+                raise errors.MissingResourceError(
+                    message=f"The resource pool with ID {resource_pool_id} or the resource "
+                    f"class with ID {class_id} do not exist, or they are not related."
+                )
+            stmt = select(schemas.NodeAffintyORM).where(schemas.NodeAffintyORM.resource_class_id == class_id)
+            res = await session.execute(stmt)
+            return [i.dump() for i in res.scalars().all()]
+
+    @_only_admins
+    async def delete_affinities(self, api_user: base_models.APIUser, resource_pool_id: int, class_id: int):
+        """Delete all affinities from a resource class."""
+        async with self.session_maker() as session, session.begin():
+            res_classes = await self.get_classes(api_user, class_id, resource_pool_id=resource_pool_id)
+            if len(res_classes) == 0:
+                raise errors.MissingResourceError(
+                    message=f"The resource pool with ID {resource_pool_id} or the resource "
+                    f"class with ID {class_id} do not exist, or they are not related."
+                )
```

Comment:
should delete be idempotent?

Reply:
Same comment as above. This is a situation where the caller is making a mistake or has a wrong idea about the state of the pools/classes that contain the toleration or affinity.